### PR TITLE
Action topic frequency

### PR DESCRIPTION
**Pull request by Nikhil-Singhal-06**
### Description
New action for checking specific topic frequency

### Pull request checklist

- [ ] [Issue](https://github.com/IntelLabs/scenario_execution/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://intellabs.github.io/scenario_execution/) reviewed and added / updated if needed (for bug fixes / features)
- [x] Format and Lint checks (`make check`) pass locally
- [x] PR applies Apache 2.0 License to all code files

### Pull request type

- [x] Feature
- [x] Documentation changes

### Current behavior
No action available to check topic frequency

### New behavior
new action assert_topic_latency

### CODE REVIEW

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+"""
+        parsed_tree = self.parser.parse_input_stream(InputStream(scenario_content))
+        model = self.parser.create_internal_model(parsed_tree, "test.osc", False)
+        scenarios = create_py_tree(model, self.parser.logger, False)
+        self.scenario_execution_ros.scenarios = scenarios
+        self.scenario_execution_ros.run()
+        self.assertTrue(self.scenario_execution_ros.process_results())
+
+
+# TODO:
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:125]: please update TODOs

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    def setUp(self) -> None:
+        rclpy.init()
+        self.running = True
+        self.parser = OpenScenario2Parser(Logger('test', False))
+        self.scenario_execution_ros = ROSScenarioExecution()
+        self.scenario_dir = get_package_share_directory('scenario_execution_ros')
+        self.received_msgs = []
+        self.node = rclpy.create_node('test_node')
+        self.publisher = self.node.create_publisher(String, "/bla", 10)
+        self.publishing_thread = threading.Thread(target=self.publish_messages, daemon=True) # TODO: use a timer
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:44]: please update TODO

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 3]
```diff
@@ -41,8 +41,9 @@ def setUp(self) -> None:
         self.received_msgs = []
         self.node = rclpy.create_node('test_node')
         self.publisher = self.node.create_publisher(String, "/bla", 10)
-        self.publishing_thread = threading.Thread(target=self.publish_messages, daemon=True) # TODO: use a timer 
-        self.publishing_thread.start()
+        self.publish_timer = threading.Timer(5, self.publish_messages)
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:44]: please use self.node.create_timer()

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        while self.running:
+            msg = String()
+            msg.data = 'Hello'
+            self.publisher.publish(msg)
+
+    def tearDown(self):
+        self.running = False
+        self.publish_timer.cancel()
+        rclpy.try_shutdown()
+
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:62]: missing tests?

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        self.node = rclpy.create_node('test_node')
+        self.publisher = self.node.create_publisher(String, "/bla", 10)
+        self.publish_timer = self.node.create_timer(5, self.publish_messages)
+        self.executor = rclpy.executors.MultiThreadedExecutor()
+        self.executor.add_node(self.node)
+        self.executor_thread = threading.Thread(target=self.executor.spin, daemon=True)
+        self.executor_thread.start()
+
+    def publish_messages(self):
+        while self.running:
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:51]: no while loop. Instead, the timer should trigger publish_message

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc [hunk 6]
```diff
@@ -107,3 +107,11 @@ action differential_drive_robot.tf_close_to:
...
     namespace: string = ''
     distance: length
+
+action assert_topic_latency:
+    topic_name: string # topic name to wait for message
+    latency: time #the time to compare against
+    comparison_operator: comparison_operator = comparison_operator!le # the comparison is done using the python operator module
+    fail_on_finish: bool = false # if true, the action should fail if comparison is true
+    rolling_average_count: int = 1 # the check is done aganist the rolling average over x elements
+    wait_for_first_message: bool = true # start measuring with the first recieved message
```
Comment (fmirus) [inline scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc:117]: @fred-labs would it make sense to add the option of allowing a certain amount of time to violate the expected latency? For instance, if the expected latency is only violated once and only by a tiny margin, the action would trigger. In some use-cases, that might not be the desired behavior. We could add something like a `tolerance_duration` (in seconds), which is by default set to 0, but could be set to a higher value when needed. What do you think?

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        result = py_trees.common.Status.FAILURE
+        if self.wait_for_first_message:
+            self.logger.info(f"Waiting for first message to publish on topic {self.topic_name}")
+            self.feedback_message = f'Waiting for first message to publish on topic {self.topic_name}'  # pylint: disable= attribute-defined-outside-init
+            result = py_trees.common.Status.RUNNING
+        else:
+            if self.msg_count > 1:
+                if self.comparison_operator(self.latency, self.average_latency) and self.fail_on_finish:
+                    result = py_trees.common.Status.FAILURE
+                    self.feedback_message = f'latency of the topic {self.topic_name} exceeds the given latency. Average latency: {self.average_latency}'  # pylint: disable= attribute-defined-outside-init
```
[the remark this answers is not in the document]
Comment (fred-labs) [reply scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:87]: yes. and please also "shrink" the message length to a minimum. E.g. remove "the topic"

**Review by fmirus — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc (hunk 6 above)
Comment (fred-labs) [reply scenario_execution_ros/scenario_execution_ros/lib_osc/ros.osc:117]: I would suggest to add that once someone asks for it. For now you could "emulate" this behavior by setting a `rolling_average_count`.

File: docs/libraries.rst [hunk 8]
```diff
@@ -184,6 +184,18 @@ Wait until a defined distance was traveled, based on odometry.
...
 - ``distance: length``: Traveled distance at which the action succeeds.
 
+``assert_topic_latency()``
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Check the latency of the specified topic (in system time). If the check with `comparison_operator` gets true, the action ends, depending on `fail_on_finish`, either with success or failure.
+
+- ``topic_name: string``:  Topic name to wait for message
+- ``latency: time``: The time to compare.
+- ``comparison_operator: comparison_operator``: operator to compare latency time. (default: ``le``)
```
[the remark this answers is not in the document]
Comment (fred-labs) [reply docs/libraries.rst:194]: Doesn't the `assert` prefix makes that clear. You normally want to assure, that a latency is below (i.e. less-or-equal) a certain threshold, right?

**Review by fred-labs — CHANGES_REQUESTED** [review]

**Review by fred-labs — CHANGES_REQUESTED** [review]

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _callback(self, msg):
+        if self.wait_for_first_message:
+            self.wait_for_first_message = False
+        else:
+            self.msg_count += 1
+            now = time.time()
+            if self.last_receive_time is not None:
+                self.latency_time = now - self.last_receive_time
+                if len(self.rolling_average_count_queue) == self.rolling_average_count:
```
[the remark this answers is not in the document]
Comment (Nikhil-Singhal-06) [reply scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:107]: But in this case, I think the average will not be based on the latest values, as we are multiplying it by previous average every time.

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        scenario_content = """
+import osc.ros
+
+scenario test_assert_topic_latency:
+    do parallel:
+        serial:
+            assert_topic_latency(
+                topic_name: '/bla',
+                latency: 0.5s,
+                wait_for_first_message: false)
```
[the remark this answers is not in the document]
Comment (fred-labs) [reply scenario_execution_ros/test/test_assert_topic_latency.py:172]: Any comment?

[the remark this answers is not in the document]
Comment (Nikhil-Singhal-06) [reply scenario_execution_ros/test/test_assert_topic_latency.py:172]: I've changed the name of the topic it subscribes to, the new topic isn't publishing any messages. 

To ensure the topic is available at the start, I've included the publisher above. This prevents the action from exiting directly with the error "No topic found."

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fmirus) [reply scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:118]: In mathematical terms, @fred-labs comment reads like this
$$\left\(\sum_{i=1}^{n-1} \frac{average_{i}}{n-1} \right\) \cdot (n-1) + \frac{average_{n}}{n} = \left\(\sum_{i=1}^{n} \frac{average_{i}}{n} \right\)$$
which is taking the latest values into account, isn't it?

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+"""
+
+        parsed_tree = self.parser.parse_input_stream(InputStream(scenario_content))
+        model = self.parser.create_internal_model(parsed_tree, "test.osc", False)
+        scenarios = create_py_tree(model, self.parser.logger, False)
+        self.scenario_execution_ros.scenarios = scenarios
+        self.scenario_execution_ros.run()
+        self.assertFalse(self.scenario_execution_ros.process_results())
+
+    def test_wait_for_first_message_false_topic_type(self):
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:265]: I would name this test  "...invalid_topic_type".

 Do you also test "wrong_topic_type"? So e.g. std_msgs.msg.Uint32

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+                topic_type: 'std_msgs.msg.ring')
+            emit end
+"""
+
+        parsed_tree = self.parser.parse_input_stream(InputStream(scenario_content))
+        model = self.parser.create_internal_model(parsed_tree, "test.osc", False)
+        scenarios = create_py_tree(model, self.parser.logger, False)
+        self.scenario_execution_ros.scenarios = scenarios
+        self.scenario_execution_ros.run()
+        self.assertFalse(self.scenario_execution_ros.process_results())
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:285]: a succeeding test with wait-for-first-message is missing.

File: scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fred-labs) [reply scenario_execution_ros/scenario_execution_ros/actions/assert_topic_latency.py:130]: any update here?

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        scenarios = create_py_tree(model, self.parser.logger, False)
+        self.scenario_execution_ros.scenarios = scenarios
+        self.scenario_execution_ros.run()
+        self.assertTrue(self.scenario_execution_ros.process_results())
+
+
+# verifies that the action throws a setup error when an incorrect 'topic_type' is set, ensuring the action's failure.
+
+
+    def test_with_false_topic_type(self):
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:114]: invalid_topic_type

File: scenario_execution_ros/test/test_assert_topic_latency.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+            emit end
+"""
+        parsed_tree = self.parser.parse_input_stream(InputStream(scenario_content))
+        model = self.parser.create_internal_model(parsed_tree, "test.osc", False)
+        scenarios = create_py_tree(model, self.parser.logger, False)
+        self.scenario_execution_ros.scenarios = scenarios
+        self.scenario_execution_ros.run()
+        self.assertFalse(self.scenario_execution_ros.process_results())
+
+
```
Comment (fred-labs) [inline scenario_execution_ros/test/test_assert_topic_latency.py:135]: Test with wrong topic type missing.

File: scenario_execution_ros/test/test_assert_topic_latency.py (hunk 12 above)
Comment (fred-labs) [reply scenario_execution_ros/test/test_assert_topic_latency.py:285]: no. with correct topic type and wait-for-first-message ==True

**Review by fred-labs — APPROVED** [review]

**Pull request merged**